pkg/policyviolation/clusterpv.go

```diff
@@ -9,144 +9,103 @@ import (
-	pv := buildPVObj(info.PolicyName, kyverno.ResourceSpec{
-		Kind: info.Resource.GetKind(),
-		Name: info.Resource.GetName(),
-	}, info.Rules,
-	)
-	return pv
+	for _, pv := range pvs {
+		// find a policy on same resource and policy combination
+		if pv.Spec.Policy == newPv.Spec.Policy &&
+			pv.Spec.ResourceSpec.Kind == newPv.Spec.ResourceSpec.Kind &&
+			pv.Spec.ResourceSpec.Name == newPv.Spec.ResourceSpec.Name {
+			return pv, nil
+		}
+	}
+	return nil, nil
 }
 
-// build PV object
-func buildPVObj(policyName string, resourceSpec kyverno.ResourceSpec, rules []kyverno.ViolatedRule) kyverno.ClusterPolicyViolation {
-	pv := kyverno.ClusterPolicyViolation{
-		Spec: kyverno.PolicyViolationSpec{
-			Policy:        policyName,
-			ResourceSpec:  resourceSpec,
-			ViolatedRules: rules,
-		},
+func (cpv *clusterPV) createPV(newPv *kyverno.ClusterPolicyViolation) error {
+	var err error
+	glog.V(4).Infof("creating new policy violation for policy %s & resource %s/%s", newPv.Spec.Policy, newPv.Spec.ResourceSpec.Kind, newPv.Spec.ResourceSpec.Name)
+	obj, err := retryGetResource(cpv.dclient, newPv.Spec.ResourceSpec)
+	if err != nil {
+		return fmt.Errorf("failed to retry getting resource for policy violation %s/%s: %v", newPv.Name, newPv.Spec.Policy, err)
 	}
+	// set owner reference to resource
+	ownerRef := createOwnerReference(obj)
+	newPv.SetOwnerReferences([]metav1.OwnerReference{ownerRef})
 
-	labelMap := map[string]string{
-		"policy":   policyName,
-		"resource": resourceSpec.ToKey(),
+	// create resource
+	_, err = cpv.kyvernoInterface.ClusterPolicyViolations().Create(newPv)
+	if err != nil {
+		glog.V(4).Infof("failed to create Cluster Policy Violation: %v", err)
+		return err
 	}
-	pv.SetLabels(labelMap)
-	pv.SetGenerateName("pv-")
-	return pv
+	glog.Infof("policy violation created for resource %v", newPv.Spec.ResourceSpec)
+	return nil
 }
 
-// build PV with owners
-func buildPVWithOwners(dclient *client.Client, info Info) []kyverno.ClusterPolicyViolation {
-	var pvs []kyverno.ClusterPolicyViolation
-	// as its blocked resource, the violation is created on owner
-	ownerMap := map[kyverno.ResourceSpec]interface{}{}
-	GetOwner(dclient, ownerMap, info.Resource)
-
-	// standaloneresource, set pvResourceSpec with resource itself
-	if len(ownerMap) == 0 {
-		pvResourceSpec := kyverno.ResourceSpec{
-			Kind: info.Resource.GetKind(),
-			Name: info.Resource.GetName(),
-		}
-		return append(pvs, buildPVObj(info.PolicyName, pvResourceSpec, info.Rules))
-	}
-
-	// Generate owner on all owners
-	for owner := range ownerMap {
-		pv := buildPVObj(info.PolicyName, owner, info.Rules)
-		pvs = append(pvs, pv)
+func (cpv *clusterPV) updatePV(newPv, oldPv *kyverno.ClusterPolicyViolation) error {
+	var err error
+	// check if there is any update
+	if reflect.DeepEqual(newPv.Spec, oldPv.Spec) {
+		glog.V(4).Infof("policy violation spec %v did not change so not updating it", newPv.Spec)
+		return nil
 	}
-	return pvs
-}
+	// set name
+	newPv.SetName(oldPv.Name)
```

Comment:
Need to SetResourceVersion here as well.

Reply:
Corrected